First working version

lukasz-gryzbon opened the pull request.

olliekennedy commented on generator/src/main/kotlin/io/lsdconsulting/stub/processor/ControllerProcessor.kt:
```diff
@@ -25,57 +34,182 @@ class ControllerProcessor : AbstractProcessor() {
+                    if (element.getAnnotation(RestController::class.java) != null) {
+                        val controllerModel = model.getControllerModel(element.toString())
+                        restControllerAnnotationHandler.handle(element, controllerModel)
+                    }
```
Could these all be like:
```
element.getAnnotation(RestController::class.java?.let {
    val controllerModel = model.getControllerModel(element.toString())
    restControllerAnnotationHandler.handle(element, controllerModel)
}
```

lukasz-gryzbon replied:
Good idea! Thanks!

olliekennedy commented on generator/src/main/kotlin/io/lsdconsulting/stub/processor/ControllerProcessor.kt:
```diff
@@ -25,57 +34,182 @@ class ControllerProcessor : AbstractProcessor() {
+
+    private fun elementsBelongingToAnnotatedClasses(annotatedClasses: List<Element>): (Element) -> Boolean =
+        {
+            annotatedClasses.contains(it) ||
```
Could use `or` instead of `||`

olliekennedy commented on generator/src/main/kotlin/io/lsdconsulting/stub/processor/PostProcessor.kt:
```diff
@@ -0,0 +1,35 @@
+    private fun updateResponseStatusOnResources(controllerModel: ControllerModel) {
+        if (controllerModel.responseStatus != null) {
+            controllerModel.resources.values.forEach {
+                if (it.responseStatus == null) it.responseStatus = controllerModel.responseStatus
```
I don't think this should be a oneliner

olliekennedy commented on generator/src/main/kotlin/io/lsdconsulting/stub/processor/typeHandlerExtensions.kt:
```diff
@@ -0,0 +1,47 @@
+}
+
+fun String.retrieveGeneric() =
+    if (this.contains("<") && this.contains(">")) {
```
I'd use `and`

olliekennedy commented on generator/src/main/kotlin/io/lsdconsulting/stub/processor/typeHandlerExtensions.kt:
```diff
@@ -0,0 +1,47 @@
+        this
+    }
+
+fun String.replacePrimitive() =
```
Love this method

olliekennedy commented on generator/src/main/kotlin/io/lsdconsulting/stub/processor/typeHandlerExtensions.kt:
Is lowercase filename the convention for files containing no classes?

lukasz-gryzbon replied:
I just follow Duncan's recommendation as I remember it... snake case is Python to me...

olliekennedy commented on generator/src/main/kotlin/io/lsdconsulting/stub/processor/typeHandlerExtensions.kt:
```diff
@@ -0,0 +1,47 @@
+        else -> this
+    }
+
+fun String.retrieveResponseType(): String? {
```
I know you've spoken before that you subscribe to the single return law, but I much prefer:
```
fun String.retrieveResponseType(): String? {
    val responseType = this.replace(Regex("\\(.*\\)"), "")
    
    if (responseType.equals("void", true)) 
        return null
    
    return responseType
}
```

Guard clauses for null suit my eye so much better than return ifs

olliekennedy commented on generator/src/main/kotlin/io/lsdconsulting/stub/writer/StubWriter.kt:
```diff
@@ -22,55 +21,28 @@ class StubWriter(processingEnv: ProcessingEnvironment) {
-            val path = Files.createFile(Path.of(stubBasePathName))
-            PrintWriter(builderFile.openWriter()).use { writer ->
-                stubBaseTemplate.evaluate(writer, mapOf("packageName" to controllerModel.packageName))
+    fun writeStubFile(model: Model) {
```
I don't like how much is going on in this method - maybe start with some newlines to tell the story better, then possibly extract methods

lukasz-gryzbon replied:
Refactored

olliekennedy commented on generator/src/test/java/io/lsdconsulting/stub/processor/ElementTypeRetrieverShould.kt:
```diff
@@ -0,0 +1,61 @@
+
+        val result = element.retrieveArgumentType()
+
+        assertThat(result, `is`("some.Class"))
```
Love how this test is structured

olliekennedy reviewed (commented):
Unbelievable stuff man! Amazing you got this together and delivering so much value already

One overall thing is I would focus on structuring your tests with more newlines to highlight underTest and your verifications.

Another thing I don't like but is completely subjective is `return if`s which feature quite a lot in this project - never been a fan but some people like them so feel free to ignore

Overall fantastic as usual - mostly style comments

lukasz-gryzbon commented on generator-integration-test/src/main/kotlin/com/lsdconsulting/stub/integration/controller/delete/DeleteRestController.kt:
```diff
@@ -0,0 +1,32 @@
+
+    @Suppress("UNUSED_PARAMETER")
+    @DeleteMapping("/resourceWithRequestBody")
+    fun resourceWithRequestBody(@RequestBody greetingRequest: GreetingRequest) {
```
Hmmm, it doesn't compile without them. Or should I put something else instead?

olliekennedy replied:
Oh sorry! It's everything but a function you can do that by the looks of it 😂 - in that case I'd say `{}` looks cleaner but up to you

lukasz-gryzbon commented on generator-integration-test/src/test/kotlin/com/lsdconsulting/stub/integration/get/GetRestControllerCustomResponseIT.kt:
How?

lukasz-gryzbon commented on generator-integration-test/src/main/kotlin/com/lsdconsulting/stub/integration/controller/delete/DeleteRestController.kt:
```diff
@@ -0,0 +1,32 @@
+        @RequestBody greetingRequest: GreetingRequest,
+        @PathVariable param: String
+    ) {
+    }
```
IntelliJ defaults to two lines 🤷

lukasz-gryzbon commented on generator-integration-test/src/main/kotlin/com/lsdconsulting/stub/integration/controller/get/GetRestController.kt:
```diff
@@ -0,0 +1,130 @@
+        @RequestParam param3: String,
+        @RequestParam param4: String
+    ) =
+        GreetingResponse(name = randomAlphabetic(10))
```
I've tried.... Hopefully you like it better

olliekennedy replied:
Better!

lukasz-gryzbon commented on generator-integration-test-api/src/main/kotlin/com/lsdconsulting/stub/integration/model/Api.kt:
```diff
@@ -0,0 +1,13 @@
+
+import com.fasterxml.jackson.annotation.JsonProperty
+
+data class GreetingResponse(
```
Good question... No meaning as far as I'm aware... 

Any better names in mind?

lukasz-gryzbon commented on generator/src/main/kotlin/io/lsdconsulting/stub/model/ControllerModel.kt:
I feel I'd like to keep the entire model in a single file...for now...

lukasz-gryzbon commented:
> Unbelievable stuff man! Amazing you got this together and delivering so much value already
> 
> One overall thing is I would focus on structuring your tests with more newlines to highlight underTest and your verifications.
> 
> Another thing I don't like but is completely subjective is `return if`s which feature quite a lot in this project - never been a fan but some people like them so feel free to ignore
> 
> Overall fantastic as usual - mostly style comments

Thanks Oli! Much appreciated! 👍 

I think I've addressed most of your comments.
